### ISSUE TITLE
Add a way for Admins to add self-signed certificates for data connections

### ISSUE THREAD

**Issue by kywalker-rh** [opendatahub-io/sig-ml-developer-experience#65]
Admins need a way to add self-signed certificates as supported  so they can use internal S3 storage instead of only public S3 storage.

See related Eng spike: https://github.com/opendatahub-io/odh-model-controller/issues/61

This is likely an admin setting, but if it is determined there is a better way to do it we should explore all options. There may be a need to show this to the user as well, and we should double check only admins will be adding the certificates, not typical users.

**Comment by lucferbux** [comment]
Mmmmm, I think this is meant to be target for every user. I feel that we might wanna have a discussion about it. Maybe @andrewballantyne can participate cause this affects to data connections too. My impression about the solution is that there would be another non-mandatory field in data connections to allow adding a ca-cert. 

The solution that @Jooho has implemented is in the data connection secret itself, so that's I think that the most straightforward translation.

But then again, that field might be weird cause it cannot be used for other components (i.e. notebook controllers and pipelines)

**Comment by kywalker-rh** [comment]
I was originally under the impression it was just for Admins, but will change this to include all users.

@andrewballantyne - By your comment I'm assuming this needs to happen after the "updated Data Connection" conversation. Is that right?

cc: @vconzola

**Comment by andrewballantyne** [comment]
Perhaps -- wouldn't hurt to have that conversation about Data Connections

**Comment by lucferbux** [comment]
> Perhaps -- wouldn't hurt to have that conversation about Data Connections

Yes, that's why I think we need to talk UI and UX wise about it. As a follow up of the work of @Jooho this seems to be directly related to data connection, but if so, it's just a patch and we need to find a proper solution (and this specific solution is problematic, cause if you add it into data connections, then users might be under the impression that this certificate could be used in workbenches, pipelines and more when it's just targeted for model serving)

**Comment by Jooho** [comment]
@lucferbux @andrewballantyne @kywalker-rh @danielezonca 
I think it can be done by global scope and also by each data connection. Definitely, admin can configure global self-signed-cert for all components(dsp, dashboard and so on) but still I think we need a way to overwrite it by end user too.  And the good place to overwrite the cert would be data connection because this is the most intuitive place, and if I were a user, I would check this object.
So why don't we use different variable names for overwriting a cert ?

This is my 2 cents:

- SELF_SIGNED_ROOT_CERT ==> for global 
- OVERWRITE_SELF_SIGNED_ROOT_CERT ==> for each component

From a component perspective, if there is no OVERWRITE_SELF_SIGNED_ROOT_CERT in dataconnection, we check if SELF_SIGNED_ROOT_CERT is avable in somewhere global CM something. If so, we convert the cert into our secret like modelmesh is doing now.

But if there is OVERWRITE_SELF_SIGNED_ROOT_CERT in dataconnection, we can use it.

The converting logic should be done by components because even kserve/modelmesh have a different way to set cert.

WDYT?

**Comment by lucferbux** [comment]
Mmmmm, I was thinking about the UX flow and what the user are expecting rather than the technical solution.
If we get this, where should users add the `SELF_SIGNED_ROOT_CERT`, in the Data Connection page? And then `OVERWRITE_SELF_SIGNED_ROOT_CERT` in the object that they want to create after selecting a Data connection?

If so, what happens if you create a new Data connection in the **"Deploy model"** modal and you add a certificate? It should be added as `SELF_SIGNED_ROOTE_CERT` or as a `OVERWRITE_SELF_SIGNED_ROOT_CERT `?

I think that solution is possible in the technical level, but we should think about the right implemetation at UI/UX level too, and there are some questions unanswered with that implementation.

**Comment by andrewballantyne** [comment]
@xianli123 is looking into this -- but he is off this week. We'll likely have to wait for him to return to looking to certs at the data connection level (for Model Serving) -- and we we'll need to look at it from the cluster level (for Elyra, & Dashboard backend service flows).

**Comment by Jooho** [comment]
As @andrewballantyne mentioned, we need to take a look at a way to set cluster-level cert. Which means it is not there for now. However, I assume that there will be a way to set cluster-level self-signed-cert somewhere finally. Based on this assumption, if I can answer your question below, it would be this.
~~~
If we get this, where should users add the SELF_SIGNED_ROOT_CERT, in the Data Connection page? And then OVERWRITE_SELF_SIGNED_ROOT_CERT in the object that they want to create after selecting a Data connection?
~~~
The cluster-level cert can be set in `SELF_SIGNED_ROOT_CERT`  in data-connection somehow. UX can show the default cert or it can be hidden.   However, if the user wants to overwrite the global cert with the data connection, it can set  `OVERWRITE_SELF_SIGNED_ROOT_CERT`. The important thing here is the environmental value is set.
From a component perspective, we can use the environmental values.
If `OVERWRITE_SELF_SIGNED_ROOT_CERT` is not set, we can use `SELF_SIGNED_ROOT_CERT`.

~~~
If so, what happens if you create a new Data connection in the "Deploy model" modal and you add a certificate? It should be added as SELF_SIGNED_ROOTE_CERT or as a OVERWRITE_SELF_SIGNED_ROOT_CERT ?
~~~
Sorry I am not familiar with this flow.  The "Deploy model" modal should be enabled after we create a data connection, no?

**Comment by lucferbux** [comment]
> Sorry I am not familiar with this flow. The "Deploy model" modal should be enabled after we create a data connection, no?

Ah no, I'm sorry, there is a scenario in which you can create the data connection while you are creating the model in the "Deploy model" modal, that's why I was asking, but if we can get UX to figure out the flow, it's just a matter of covering all the edge cases in the implemntation, thanks!
